L1Trigger/TrackerTFP/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,207 @@
+    edPutTokenLostStubs_ = produces<TTDTC::Streams>(branchLostStubs);
+    edPutTokenLostTracks_ = produces<StreamsTrack>(branchLostTracks);
+    // book ES products
+    esGetTokenSetup_ = esConsumes<Setup, SetupRcd, Transition::BeginRun>();
+    esGetTokenDataFormats_ = esConsumes<DataFormats, DataFormatsRcd, Transition::BeginRun>();
+    esGetTokenLayerEncoding_ = esConsumes<LayerEncoding, LayerEncodingRcd, Transition::BeginRun>();
+    // initial ES products
+    setup_ = nullptr;
+    dataFormats_ = nullptr;
+    layerEncoding_ = nullptr;
+    //
+    enableTruncation_ = iConfig.getParameter<bool>("EnableTruncation");
+  }
+
+  void ProducerKFin::beginRun(const Run& iRun, const EventSetup& iSetup) {
+    // helper class to store configurations
+    setup_ = &iSetup.getData(esGetTokenSetup_);
+    if (!setup_->configurationSupported())
+      return;
+    // check process history if desired
+    if (iConfig_.getParameter<bool>("CheckHistory"))
+      setup_->checkHistory(iRun.processHistory());
+    // helper class to extract structured data from TTDTC::Frames
+    dataFormats_ = &iSetup.getData(esGetTokenDataFormats_);
+    // helper class to encode layer
+    layerEncoding_ = &iSetup.getData(esGetTokenLayerEncoding_);
+  }
+
+  void ProducerKFin::produce(Event& iEvent, const EventSetup& iSetup) {
+    const DataFormat& dfCot = dataFormats_->format(Variable::cot, Process::sf);
+    const DataFormat& dfZT = dataFormats_->format(Variable::zT, Process::sf);
+    // empty KFin products
+    TTDTC::Streams streamAcceptedStubs(dataFormats_->numStreams(Process::kf) * setup_->numLayers());
+    StreamsTrack streamAcceptedTracks(dataFormats_->numStreams(Process::kf));
+    TTDTC::Streams streamLostStubs(dataFormats_->numStreams(Process::kf) * setup_->numLayers());
+    StreamsTrack streamLostTracks(dataFormats_->numStreams(Process::kf));
+    // read in SFout Product and produce KFin product
+    if (setup_->configurationSupported()) {
+      Handle<TTDTC::Streams> handleStubs;
+      iEvent.getByToken<TTDTC::Streams>(edGetTokenStubs_, handleStubs);
+      const TTDTC::Streams& streams = *handleStubs.product();
+      Handle<vector<TTTrack<Ref_Phase2TrackerDigi_>>> handleTTTracks;
+      iEvent.getByToken<vector<TTTrack<Ref_Phase2TrackerDigi_>>>(edGetTokenTTTracks_, handleTTTracks);
+      const vector<TTTrack<Ref_Phase2TrackerDigi_>>& ttTracks = *handleTTTracks.product();
+      for (int region = 0; region < setup_->numRegions(); region++) {
+        int nStubsSF(0);
+        for (int channel = 0; channel < dataFormats_->numChannel(Process::sf); channel++) {
+          const int index = region * dataFormats_->numChannel(Process::sf) + channel;
+          const TTDTC::Stream& stream = streams[index];
+          nStubsSF += accumulate(stream.begin(), stream.end(), 0, [](int& sum, const TTDTC::Frame& frame){ return sum += frame.first.isNonnull() ? 1 : 0; });
+        }
+        vector<StubSF> stubsSF;
+        stubsSF.reserve(nStubsSF);
```

Comment:
The count nStubsSF is needed only to reserve space in the vector stubsSF. Could this code be deleted if the vector were replaced by an std::list (which needs no advanced space reservation)? I see nothing that requires the use of a vector.

Reply:
vectors are faster then lists and therefor superior if size is static. instead of seeing it like I don't have to reserve a list I see it like I can't reserve a list which is a drawback.

Reply:
If code optimisation adds significantly to code complexity (and the calc of nStubsSF is complex), then it should only be done if the CPU profiler indicates there is a problem. Keeping code simple is also essential. Otherwise it will be challenging for CMS to maintain it.